File: scripts/fixDictionary.js
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

// Common English words with proper Traditional Chinese definitions and sentences
// Suitable for Taiwanese spelling bee competition
const WORD_DEFINITIONS = {
  // Elementary level (A1)
  "say": {
    definition: "說，講",
    sentence: "Please say your name clearly.",
    partOfSpeech: "動詞"
  },
  "get": {
    definition: "得到，獲得",
    sentence: "I want to get good grades in school.",
    partOfSpeech: "動詞"
  },
  "saw": {
    definition: "看見（see的過去式）",
    sentence: "I saw a beautiful bird yesterday.",
    partOfSpeech: "動詞"
  },
  "use": {
    definition: "使用，利用",
    sentence: "We use computers every day.",
    partOfSpeech: "動詞"
  },
  "way": {
    definition: "方法，道路",
    sentence: "This is the best way to learn English.",
    partOfSpeech: "名詞"
  },
  "new": {
    definition: "新的，嶄新的",
    sentence: "I got a new backpack for school.",
    partOfSpeech: "形容詞"
  },
  "old": {
    definition: "老的，舊的",
    sentence: "My grandfather is very old and wise.",
    partOfSpeech: "形容詞"
  },
  "see": {
    definition: "看見，看到",
    sentence: "I can see the mountains from my window.",
    partOfSpeech: "動詞"
  },
  "make": {
    definition: "製作，做",
    sentence: "Let's make a cake for the party.",
    partOfSpeech: "動詞"
  },
  "come": {
    definition: "來，過來",
    sentence: "Please come to my birthday party.",
    partOfSpeech: "動詞"
  },
  "look": {
    definition: "看，注視",
    sentence: "Look at the beautiful sunset!",
    partOfSpeech: "動詞"
  },
  "find": {
    definition: "找到，發現",
    sentence: "I can't find my pencil case.",
    partOfSpeech: "動詞"
  },
  "give": {
    definition: "給，給予",
    sentence: "Please give me your homework.",
    partOfSpeech: "動詞"
  },
  "tell": {
    definition: "告訴，講述",
    sentence: "Can you tell me the time?",
    partOfSpeech: "動詞"
  },
  "work": {
    definition: "工作，運作",
    sentence: "My father works at a hospital.",
    partOfSpeech: "動詞"
  },
  "call": {
    definition: "打電話，叫",
    sentence: "I will call you tomorrow.",
    partOfSpeech: "動詞"
  },
  "feel": {
    definition: "感覺，覺得",
    sentence: "I feel happy when I see my friends.",
    partOfSpeech: "動詞"
  },
  "hand": {
    definition: "手",
    sentence: "Please raise your hand if you know the answer.",
    partOfSpeech: "名詞"
  },
  "high": {
    definition: "高的",
    sentence: "The mountain is very high.",
    partOfSpeech: "形容詞"
  },
  "year": {
    definition: "年，年度",
    sentence: "I am ten years old this year.",
    partOfSpeech: "名詞"
  },
  "different": {
    definition: "不同的，相異的",
    sentence: "People have different hobbies.",
    partOfSpeech: "形容詞"
  },
  "important": {
    definition: "重要的",
    sentence: "It is important to study hard.",
    partOfSpeech: "形容詞"
  },
  "beautiful": {
    definition: "美麗的，漂亮的",
    sentence: "Taiwan has many beautiful landscapes.",
    partOfSpeech: "形容詞"
  },
  "wonderful": {
    definition: "精彩的，極好的",
    sentence: "We had a wonderful time at the museum.",
    partOfSpeech: "形容詞"
  },
  "because": {
    definition: "因為",
    sentence: "I study English because it is useful.",
    partOfSpeech: "連接詞"
  }
}

// Generate definition and sentence for a word
function generateContent(word, cefr, originalPartOfSpeech) {
  // Check if we have a manual definition
  if (WORD_DEFINITIONS[word]) {
    return WORD_DEFINITIONS[word]
  }

  // For words without manual definitions, generate based on patterns
  // This is a simplified version - ideally you'd use an API or manual curation

  const partOfSpeech = originalPartOfSpeech || "名詞"

  // Generate basic definition (you'll need to replace these with real definitions)
  const definition = `${word}（待補充翻譯）`

  // Generate simple sentence based on part of speech
  let sentence
  if (partOfSpeech === "動詞") {
    sentence = `Students learn to ${word} in class.`
  } else if (partOfSpeech === "形容詞") {
    sentence = `The book is very ${word}.`
  } else if (partOfSpeech === "名詞") {
    sentence = `The ${word} is on the table.`
  } else if (partOfSpeech === "副詞") {
    sentence = `She speaks ${word}.`
  } else {
    sentence = `This sentence uses the word "${word}".`
  }

  return { definition, sentence, partOfSpeech }
}

// Process a dictionary file
function processDictionary(filename) {
  const filePath = path.join(__dirname, '..', 'src', 'gameTypes', 'english-spelling', 'data', filename)

  console.log(`\nProcessing ${filename}...`)

  const data = JSON.parse(fs.readFileSync(filePath, 'utf8'))
  let fixed = 0
  let needsManual = 0

  const updated = data.map(entry => {
    // Check if entry needs fixing
    const needsFix = entry.definition.includes('的中文意思') ||
                     entry.sentence.includes('was found') ||
                     entry.sentence.includes('is useful') ||
                     entry.sentence.includes('must ') ||
                     entry.sentence.includes('We have a ')

    if (needsFix) {
      const content = generateContent(entry.word, entry.cefr, entry.partOfSpeech)

      if (content.definition.includes('待補充翻譯')) {
        needsManual++
      } else {
        fixed++
      }

      return {
        ...entry,
        definition: content.definition,
        sentence: content.sentence,
        partOfSpeech: content.partOfSpeech
      }
    }

    return entry
  })

  // Save updated file
  fs.writeFileSync(filePath, JSON.stringify(updated, null, 2), 'utf8')

  console.log(`  ✅ Fixed: ${fixed} entries`)
  console.log(`  ⚠️  Needs manual review: ${needsManual} entries`)
  console.log(`  Total: ${data.length} entries`)

  return { fixed, needsManual, total: data.length }
}

// Main execution
console.log('🔧 Dictionary Fixer')
console.log('===================\n')
console.log('NOTE: This script fixes obvious placeholder text.')
console.log('Words without manual definitions will be marked for review.\n')

const files = [
  'elementary.json',
  'middle.json',
  'high.json',
  'university.json',
  'expert.json'
]

let totalFixed = 0
let totalNeedsManual = 0
let totalWords = 0

files.forEach(file => {
  const result = processDictionary(file)
  totalFixed += result.fixed
  totalNeedsManual += result.needsManual
  totalWords += result.total
})

console.log('\n' + '='.repeat(50))
console.log('\nSummary:')
console.log(`  Total words: ${totalWords}`)
console.log(`  Fixed with proper definitions: ${totalFixed}`)
console.log(`  Needs manual translation: ${totalNeedsManual}`)
console.log(`\nNext steps:`)
console.log(`  1. Review files marked "待補充翻譯"`)
console.log(`  2. Add more entries to WORD_DEFINITIONS in this script`)
console.log(`  3. Re-run the script to apply more fixes`)
console.log(`  4. Consider using translation API for remaining words\n`)

File: fix_dictionary.py
#!/usr/bin/env python3
"""
Script to fix all dictionary entries in the spelling bee game.
Corrects definitions, sentences, and parts of speech.
"""

import json
import os

# Comprehensive word corrections database
WORD_CORRECTIONS = {
    # Elementary level common words
    "say": {
        "definition": "說",
        "sentence": "I want to say thank you for your help.",
        "partOfSpeech": "動詞"
    },
    "get": {
        "definition": "得到、獲得",
        "sentence": "I need to get some milk from the store.",
        "partOfSpeech": "動詞"
    },
    "saw": {
        "definition": "看見（see的過去式）",
        "sentence": "I saw a beautiful bird in the garden yesterday.",
        "partOfSpeech": "動詞"
    },
    "use": {
        "definition": "使用",
        "sentence": "Please use a pencil to write your name.",
        "partOfSpeech": "動詞"
    },
    "ask": {
        "definition": "詢問、請求",
        "sentence": "You can ask the teacher if you have questions.",
        "partOfSpeech": "動詞"
    },
    "put": {
        "definition": "放置",
        "sentence": "Please put your books on the table.",
        "partOfSpeech": "動詞"
    },
    "see": {
        "definition": "看見",
        "sentence": "I can see the mountains from my window.",
        "partOfSpeech": "動詞"
    },
    "run": {
        "definition": "跑步",
        "sentence": "The children run in the playground every day.",
        "partOfSpeech": "動詞"
    },
    "come": {
        "definition": "來",
        "sentence": "Please come to my birthday party on Saturday.",
        "partOfSpeech": "動詞"
    },
    "make": {
        "definition": "製作、使得",
        "sentence": "My mother can make delicious cookies.",
        "partOfSpeech": "動詞"
    },
    "know": {
        "definition": "知道",
        "sentence": "I know the answer to this question.",
        "partOfSpeech": "動詞"
    },
    "take": {
        "definition": "拿、帶",
        "sentence": "Don't forget to take your umbrella today.",
        "partOfSpeech": "動詞"
    },
    "find": {
        "definition": "找到",
        "sentence": "I can't find my pencil case anywhere.",
        "partOfSpeech": "動詞"
    },
    "give": {
        "definition": "給予",
        "sentence": "Please give me a piece of paper.",
        "partOfSpeech": "動詞"
    },
    "tell": {
        "definition": "告訴",
        "sentence": "Can you tell me the time, please?",
        "partOfSpeech": "動詞"
    },
    "work": {
        "definition": "工作",
        "sentence": "My father works in an office downtown.",
        "partOfSpeech": "動詞"
    },
    "call": {
        "definition": "呼叫、打電話",
        "sentence": "I will call you when I arrive home.",
        "partOfSpeech": "動詞"
    },
    "try": {
        "definition": "嘗試",
        "sentence": "You should try your best on the test.",
        "partOfSpeech": "動詞"
    },
    "feel": {
        "definition": "感覺",
        "sentence": "I feel happy when I play with my friends.",
        "partOfSpeech": "動詞"
    },
    "leave": {
        "definition": "離開",
        "sentence": "We need to leave for school at 7:30.",
        "partOfSpeech": "動詞"
    },
    "keep": {
        "definition": "保持、保存",
        "sentence": "Please keep your room clean and tidy.",
        "partOfSpeech": "動詞"
    },
    "let": {
        "definition": "讓",
        "sentence": "Let me help you carry those heavy bags.",
        "partOfSpeech": "動詞"
    },
    "begin": {
        "definition": "開始",
        "sentence": "The movie will begin at eight o'clock.",
        "partOfSpeech": "動詞"
    },
    "seem": {
        "definition": "似乎",
        "sentence": "You seem tired today. Are you okay?",
        "partOfSpeech": "動詞"
    },
    "help": {
        "definition": "幫助",
        "sentence": "Can you help me with my homework?",
        "partOfSpeech": "動詞"
    },
    "show": {
        "definition": "展示、顯示",
        "sentence": "Let me show you my new toy.",
        "partOfSpeech": "動詞"
    },
    "hear": {
        "definition": "聽見",
        "sentence": "I can hear music coming from next door.",
        "partOfSpeech": "動詞"
    },
    "play": {
        "definition": "玩、演奏",
        "sentence": "The children play basketball after school.",
        "partOfSpeech": "動詞"
    },
    "move": {
        "definition": "移動",
        "sentence": "We will move to a new house next month.",
        "partOfSpeech": "動詞"
    },
    "live": {
        "definition": "居住",
        "sentence": "I live in a small apartment with my family.",
        "partOfSpeech": "動詞"
    },
    "believe": {
        "definition": "相信",
        "sentence": "I believe you can do well on your test.",
        "partOfSpeech": "動詞"
    },
    "bring": {
        "definition": "帶來",
        "sentence": "Please bring your textbook to class tomorrow.",
        "partOfSpeech": "動詞"
    },
    "happen": {
        "definition": "發生",
        "sentence": "What will happen if it rains tomorrow?",
        "partOfSpeech": "動詞"
    },
    "write": {
        "definition": "寫",
        "sentence": "I write in my diary every night before bed.",
        "partOfSpeech": "動詞"
    },
    "sit": {
        "definition": "坐",
        "sentence": "Please sit down and make yourself comfortable.",
        "partOfSpeech": "動詞"
    },
    "stand": {
        "definition": "站立",
        "sentence": "Everyone must stand when the principal enters.",
        "partOfSpeech": "動詞"
    },
    "lose": {
        "definition": "失去、輸掉",
        "sentence": "I hope we don't lose the game today.",
        "partOfSpeech": "動詞"
    },
    "pay": {
        "definition": "付款",
        "sentence": "You need to pay for the books at the counter.",
        "partOfSpeech": "動詞"
    },
    "meet": {
        "definition": "遇見、會面",
        "sentence": "I will meet my friends at the library.",
        "partOfSpeech": "動詞"
    },
    "include": {
        "definition": "包括",
        "sentence": "The price includes breakfast and lunch.",
        "partOfSpeech": "動詞"
    },
    "continue": {
        "definition": "繼續",
        "sentence": "Please continue reading until I tell you to stop.",
        "partOfSpeech": "動詞"
    },
    "set": {
        "definition": "設置、放置",
        "sentence": "Please set the table for dinner.",
        "partOfSpeech": "動詞"
    },
    "learn": {
        "definition": "學習",
        "sentence": "I want to learn how to play the piano.",
        "partOfSpeech": "動詞"
    },
    "change": {
        "definition": "改變",
        "sentence": "People can change their habits with practice.",
        "partOfSpeech": "動詞"
    },
    "lead": {
        "definition": "領導、引導",
        "sentence": "The teacher will lead us to the cafeteria.",
        "partOfSpeech": "動詞"
    },
    "understand": {
        "definition": "理解",
        "sentence": "Do you understand the instructions?",
        "partOfSpeech": "動詞"
    },
    "watch": {
        "definition": "觀看",
        "sentence": "I like to watch cartoons on Saturday morning.",
        "partOfSpeech": "動詞"
    },
    "follow": {
        "definition": "跟隨",
        "sentence": "Please follow me to the classroom.",
        "partOfSpeech": "動詞"
    },
    "stop": {
        "definition": "停止",
        "sentence": "The bus will stop at the next corner.",
        "partOfSpeech": "動詞"
    },
    "create": {
        "definition": "創造",
        "sentence": "The artist can create beautiful paintings.",
        "partOfSpeech": "動詞"
    },
    "speak": {
        "definition": "說話",
        "sentence": "Please speak louder so everyone can hear you.",
        "partOfSpeech": "動詞"
    },
}

# Part of speech translation mapping
POS_MAPPING = {
    "noun": "名詞",
    "verb": "動詞",
    "adjective": "形容詞",
    "adverb": "副詞",
    "pronoun": "代名詞",
    "preposition": "介係詞",
    "conjunction": "連接詞",
    "interjection": "感嘆詞"
}

def infer_pos_from_word(word):
    """Infer part of speech from common patterns"""
    if word.endswith('ly'):
        return "副詞"
    if word.endswith('ness') or word.endswith('tion') or word.endswith('ment'):
        return "名詞"
    if word.endswith('ful') or word.endswith('less') or word.endswith('ous') or word.endswith('ive'):
        return "形容詞"
    # Default to verb for simple words
    return "動詞"

def generate_definition(word, pos_eng):
    """Generate a definition based on word and part of speech"""
    # This would ideally use a dictionary API, but for now we'll use patterns
    if word in WORD_CORRECTIONS:
        return WORD_CORRECTIONS[word]["definition"]

    # Fallback: keep existing if not placeholder
    return f"{word}（{POS_MAPPING.get(pos_eng, '詞')}）"

def generate_sentence(word, pos_eng):
    """Generate a grammatically correct sentence"""
    if word in WORD_CORRECTIONS:
        return WORD_CORRECTIONS[word]["sentence"]

    # Generate based on part of speech
    if pos_eng == "verb" or pos_eng == "動詞":
        return f"I {word} every day."
    elif pos_eng == "noun" or pos_eng == "名詞":
        return f"The {word} is very important."
    elif pos_eng == "adjective" or pos_eng == "形容詞":
        return f"This is a {word} example."
    elif pos_eng == "adverb" or pos_eng == "副詞":
        return f"She worked {word} on the project."
    else:
        return f"This is an example using the word {word}."

def correct_pos(word, current_pos):
    """Correct the part of speech"""
    if word in WORD_CORRECTIONS:
        return WORD_CORRECTIONS[word]["partOfSpeech"]

    # If already in Chinese, try to validate it
    if current_pos in ["名詞", "動詞", "形容詞", "副詞", "代名詞", "介係詞", "連接詞", "感嘆詞"]:
        return current_pos

    # Convert from English if needed
    if current_pos.lower() in POS_MAPPING:
        return POS_MAPPING[current_pos.lower()]

    # Infer from word
    return infer_pos_from_word(word)

def fix_entry(entry):
    """Fix a single word entry"""
    word = entry["word"]
    definition = entry.get("definition", "")
    sentence = entry.get("sentence", "")
    pos = entry.get("partOfSpeech", "")

    # Fix definition if it's a placeholder
    if "的中文意思" in definition or not definition:
        entry["definition"] = generate_definition(word, pos)

    # Fix sentence if it's grammatically incorrect or too generic
    if (
        sentence.startswith("The " + word + " was found") or
        sentence.startswith("The " + word + " is important") or
        sentence.startswith("This " + word + " is useful") or
        sentence.startswith("I need a " + word) or
        sentence.startswith("We must " + word) or
        "was found" in sentence or
        not sentence
    ):
        entry["sentence"] = generate_sentence(word, pos)

    # Fix part of speech
    entry["partOfSpeech"] = correct_pos(word, pos)

    return entry

def process_file(filepath):
    """Process a single JSON file"""
    print(f"Processing {filepath}...")

    with open(filepath, 'r', encoding='utf-8') as f:
        data = json.load(f)

    # Fix each entry
    for i, entry in enumerate(data):
        data[i] = fix_entry(entry)
        if (i + 1) % 100 == 0:
            print(f"  Processed {i + 1}/{len(data)} entries...")

    # Write back with proper formatting
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

    print(f"✓ Completed {filepath} - {len(data)} entries fixed\n")

def main():
    base_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "src", "gameTypes", "english-spelling", "data")

    files = [
        "elementary.json",
        "middle.json",
        "high.json",
        "university.json",
        "expert.json"
    ]

    for filename in files:
        filepath = os.path.join(base_path, filename)
        if os.path.exists(filepath):
            process_file(filepath)
        else:
            print(f"Warning: {filepath} not found")

if __name__ == "__main__":
    main()
